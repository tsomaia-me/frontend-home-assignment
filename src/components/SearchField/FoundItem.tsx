import React, { useCallback } from 'react'

export interface FoundItemProps {
  packageName: string
  onClick: (item: string) => void
}

function FoundItem({ packageName, onClick }: FoundItemProps) {
  const handleClick = useCallback(() => onClick(packageName), [packageName, onClick])

  return (
    <li onMouseDown={handleItemMouseDown} onClick={handleClick}>
      {packageName}
    </li>
  )
}

function handleItemMouseDown(event: React.MouseEvent) {
  event.preventDefault()
}

export default FoundItem
