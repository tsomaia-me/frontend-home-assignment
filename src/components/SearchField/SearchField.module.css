.root {
    position: relative;
    height: 40px;
}

.container {
    width: 100%;
    position: absolute;
    background: white;
    border: 1px solid #e0e0e0;
    border-radius: 8px;
}

.container:focus-within {
    box-shadow: 2px 4px 4px rgba(0, 0, 0, 0.32);
}

.container > input {
    background: transparent;
    border: none;
}

.list {
    margin: 0;
    padding: 0;
    list-style-type: none;
    max-height: 0;
    overflow-y: auto;
    border-bottom-left-radius: 8px;
    border-bottom-right-radius: 8px;
}

.list > li {
    padding: 8px 16px;
    cursor: default;
}

.list > li:hover, .list > li.active {
    background-color: #e0e0e0;
}

.container:focus-within > .list {
    max-height: 300px;
}
