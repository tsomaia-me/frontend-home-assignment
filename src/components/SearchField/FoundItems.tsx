import React from 'react'
import { observer } from 'mobx-react'
import { usePackageStore } from '../../hooks/usePackageStore'
import styles from './SearchField.module.css'
import FoundItem from './FoundItem'

export interface FoundItemsProps {
  onSelect: (packageName: string) => void
}

function FoundItems({ onSelect }: FoundItemsProps) {
  const packageStore = usePackageStore()

  return (
    <ul className={styles.list}>
      {packageStore.foundPackages.map(packageName => (
        <FoundItem
          key={packageName}
          packageName={packageName}
          onClick={onSelect}
        />
      ))}
    </ul>
  )
}

export default observer(FoundItems)
