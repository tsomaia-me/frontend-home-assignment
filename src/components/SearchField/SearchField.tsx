import React, { ChangeEvent, useCallback, useEffect, useState } from 'react'
import { observer } from 'mobx-react'

import styles from './SearchField.module.css'
import { useDebouncedValue, useEnvironment } from '../../hooks'
import { usePackageStore } from '../../hooks/usePackageStore'
import FoundItems from './FoundItems'
import Input from '../Input/Input'

function SearchField() {
  const { debounceTimeout } = useEnvironment()
  const packageStore = usePackageStore()
  const [input, setInput] = useState(packageStore.selectedPackageName ?? '')
  const debouncedInput = useDebouncedValue(input, debounceTimeout)
  const handleInputChange = useCallback((e: ChangeEvent) => setInput((e.target as HTMLInputElement).value), [])
  const selectPackage = useCallback((packageName: string) => {
    (document.activeElement as HTMLInputElement | null)?.blur()
    window.history.pushState({}, '', packageName)
    packageStore.selectPackage(packageName)
  }, [packageStore])
  const handleKeyPress = useCallback((event: React.KeyboardEvent) => {
    if (event.key === 'Enter') {
      selectPackage(input)
    }
  }, [input, selectPackage])

  useEffect(() => {
    packageStore.searchPackages(debouncedInput)
  }, [debouncedInput, packageStore])

  return (
    <div className={styles.root}>
      <div className={styles.container}>
        <Input
          type="search"
          placeholder="Search packages..."
          value={input}
          onChange={handleInputChange}
          onKeyPress={handleKeyPress}
        />

        <FoundItems onSelect={selectPackage}/>
      </div>
    </div>
  )
}

export default observer(SearchField)
