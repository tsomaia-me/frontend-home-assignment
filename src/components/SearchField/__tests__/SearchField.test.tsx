import React from 'react'
import renderer from 'react-test-renderer'
import SearchField from '../SearchField'

test('renders SearchField correctly', () => {
  const json = renderer.create(<SearchField/>).toJSON()
  expect(json).toMatchSnapshot()
})
