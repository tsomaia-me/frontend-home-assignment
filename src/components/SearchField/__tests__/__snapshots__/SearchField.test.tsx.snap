// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SearchField correctly 1`] = `
<div
  className="root"
>
  <div
    className="container"
  >
    <input
      className="root"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Search packages..."
      type="search"
      value=""
    />
    <ul
      className="list"
    />
  </div>
</div>
`;
